File: examples/csharp/google-docs/GoogleDocsExample.Tests/DriveTest.cs
using System.Net;
using System.Net.Http.Json;
using Authzed.Api.V1;
using Borkfork.SpiceDb.Examples.GoogleDocs;
using Xunit;

namespace GoogleDocsExample.Tests;

public class DriveTest
{
    private static Relationship Rel(string resource, string relation, string subject)
    {
        var resParts = resource.Split(':');
        var subParts = subject.Split(':');
        return new Relationship
        {
            Resource = new ObjectReference { ObjectType = resParts[0], ObjectId = resParts[1] },
            Relation = relation,
            Subject = new SubjectReference
            {
                Object = new ObjectReference { ObjectType = subParts[0], ObjectId = subParts[1] }
            }
        };
    }

    [Fact]
    public async Task Grants_alice_viewer_on_folder_and_verifies_view_on_documents()
    {
        var (app, spicedb) = Server.CreateApp(Array.Empty<Relationship>());
        using (spicedb)
        {
            _ = spicedb.Permissions().WriteRelationships(new WriteRelationshipsRequest
            {
                Updates =
                {
                    new RelationshipUpdate { Operation = RelationshipUpdate.Types.Operation.Touch, Relationship = Rel("document:doc1", "folder", "folder:marketing") },
                    new RelationshipUpdate { Operation = RelationshipUpdate.Types.Operation.Touch, Relationship = Rel("document:doc2", "folder", "folder:marketing") },
                }
            });
            _ = spicedb.Permissions().WriteRelationships(new WriteRelationshipsRequest
            {
                Updates = { new RelationshipUpdate { Operation = RelationshipUpdate.Types.Operation.Touch, Relationship = Rel("folder:marketing", "viewer", "user:alice") } }
            });

            var port = 30000 + Random.Shared.Next(0, 10000);
            app.Urls.Add($"http://127.0.0.1:{port}");
            await app.StartAsync();
            try
            {
                using var client = new HttpClient { BaseAddress = new Uri($"http://127.0.0.1:{port}") };
                client.DefaultRequestHeaders.Add("X-User", "alice");

                var res1 = await client.GetAsync("/drive/marketing/doc1");
                var res2 = await client.GetAsync("/drive/marketing/doc2");

                Assert.Equal(HttpStatusCode.OK, res1.StatusCode);
                var body1 = await res1.Content.ReadFromJsonAsync<DriveResponse>();
                Assert.NotNull(body1);
                Assert.True(body1.Ok);
                Assert.Equal("alice", body1.User);
                Assert.Equal("doc1", body1.Document);

                Assert.Equal(HttpStatusCode.OK, res2.StatusCode);
                var body2 = await res2.Content.ReadFromJsonAsync<DriveResponse>();
                Assert.NotNull(body2);
                Assert.True(body2.Ok);
                Assert.Equal("doc2", body2.Document);
            }
            finally
            {
                await app.StopAsync();
            }
        }
    }

    [Fact]
    public async Task Denies_bob_when_not_viewer()
    {
        var (app, spicedb) = Server.CreateApp(new[]
        {
            Rel("document:doc1", "folder", "folder:marketing"),
            Rel("folder:marketing", "viewer", "user:alice")
        });
        using (spicedb)
        {
            var port = 30000 + Random.Shared.Next(0, 10000);
            app.Urls.Add($"http://127.0.0.1:{port}");
            await app.StartAsync();
            try
            {
                using var client = new HttpClient { BaseAddress = new Uri($"http://127.0.0.1:{port}") };
                client.DefaultRequestHeaders.Add("X-User", "bob");

                var res = await client.GetAsync("/drive/marketing/doc1");

                Assert.Equal(HttpStatusCode.Forbidden, res.StatusCode);
                var body = await res.Content.ReadFromJsonAsync<ErrorResponse>();
                Assert.NotNull(body);
                Assert.Equal("Forbidden", body.Error);
            }
            finally
            {
                await app.StopAsync();
            }
        }
    }

    [Fact]
    public async Task Inherits_view_from_parent_folder()
    {
        var (app, spicedb) = Server.CreateApp(new[]
        {
            Rel("folder:marketing", "parent", "folder:root"),
            Rel("document:doc1", "folder", "folder:marketing"),
            Rel("folder:root", "viewer", "user:alice")
        });
        using (spicedb)
        {
            var port = 30000 + Random.Shared.Next(0, 10000);
            app.Urls.Add($"http://127.0.0.1:{port}");
            await app.StartAsync();
            try
            {
                using var client = new HttpClient { BaseAddress = new Uri($"http://127.0.0.1:{port}") };
                client.DefaultRequestHeaders.Add("X-User", "alice");

                var res = await client.GetAsync("/drive/marketing/doc1");

                Assert.Equal(HttpStatusCode.OK, res.StatusCode);
                var body = await res.Content.ReadFromJsonAsync<DriveResponse>();
                Assert.NotNull(body);
                Assert.True(body.Ok);
                Assert.Equal("alice", body.User);
                Assert.Equal("doc1", body.Document);
            }
            finally
            {
                await app.StopAsync();
            }
        }
    }

    [Fact]
    public async Task User_can_be_granted_permission_directly_on_document()
    {
        var (app, spicedb) = Server.CreateApp(new[]
        {
            Rel("document:doc1", "folder", "folder:marketing"),
            Rel("document:doc1", "viewer", "user:alice")
        });
        using (spicedb)
        {
            var port = 30000 + Random.Shared.Next(0, 10000);
            app.Urls.Add($"http://127.0.0.1:{port}");
            await app.StartAsync();
            try
            {
                using var client = new HttpClient { BaseAddress = new Uri($"http://127.0.0.1:{port}") };
                client.DefaultRequestHeaders.Add("X-User", "alice");

                var res = await client.GetAsync("/drive/marketing/doc1");

                Assert.Equal(HttpStatusCode.OK, res.StatusCode);
                var body = await res.Content.ReadFromJsonAsync<DriveResponse>();
                Assert.NotNull(body);
                Assert.True(body.Ok);
                Assert.Equal("doc1", body.Document);
            }
            finally
            {
                await app.StopAsync();
            }
        }
    }

    [Fact]
    public async Task Returns_401_when_X_User_header_missing()
    {
        var (app, spicedb) = Server.CreateApp(new[]
        {
            Rel("document:doc1", "folder", "folder:marketing"),
            Rel("folder:marketing", "viewer", "user:alice")
        });
        using (spicedb)
        {
            var port = 30000 + Random.Shared.Next(0, 10000);
            app.Urls.Add($"http://127.0.0.1:{port}");
            await app.StartAsync();
            try
            {
                using var client = new HttpClient { BaseAddress = new Uri($"http://127.0.0.1:{port}") };

                var res = await client.GetAsync("/drive/marketing/doc1");

                Assert.Equal(HttpStatusCode.Unauthorized, res.StatusCode);
                var body = await res.Content.ReadFromJsonAsync<ErrorResponse>();
                Assert.NotNull(body);
                Assert.Equal("Missing X-User header", body.Error);
            }
            finally
            {
                await app.StopAsync();
            }
        }
    }

    private sealed class DriveResponse
    {
        public bool Ok { get; set; }
        public string? User { get; set; }
        public string? Folder { get; set; }
        public string? Document { get; set; }
        public string? Message { get; set; }
    }

    private sealed class ErrorResponse
    {
        public string? Error { get; set; }
        public string? Message { get; set; }
    }
}

File: examples/java/google-docs/src/test/java/com/borkfork/spicedb/examples/googledocs/DriveTest.java
package com.borkfork.spicedb.examples.googledocs;

import static org.junit.jupiter.api.Assertions.*;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import org.junit.jupiter.api.Test;

class DriveTest {

  private static HttpResponse<String> get(
      int port, String path, String xUser) throws Exception {
    var client = HttpClient.newHttpClient();
    var builder =
        HttpRequest.newBuilder()
            .uri(URI.create("http://127.0.0.1:" + port + path))
            .GET();
    if (xUser != null) {
      builder.header("X-User", xUser);
    }
    return client.send(builder.build(), HttpResponse.BodyHandlers.ofString());
  }

  private static JsonObject json(String body) {
    return new Gson().fromJson(body, JsonObject.class);
  }

  @Test
  void aliceViewerCanViewDocuments() throws Exception {
    var appAndDb =
        Server.createApp(
            List.of(
                Server.rel("document:doc1", "folder", "folder:marketing"),
                Server.rel("document:doc2", "folder", "folder:marketing"),
                Server.rel("folder:marketing", "viewer", "user:alice")));

    appAndDb.app().start(0);
    int port = appAndDb.app().port();

    try {
      var r1 = get(port, "/drive/marketing/doc1", "alice");
      var r2 = get(port, "/drive/marketing/doc2", "alice");

      assertEquals(200, r1.statusCode());
      assertTrue(json(r1.body()).get("ok").getAsBoolean());
      assertEquals("alice", json(r1.body()).get("user").getAsString());
      assertEquals("doc1", json(r1.body()).get("document").getAsString());
      assertEquals(200, r2.statusCode());
      assertEquals("doc2", json(r2.body()).get("document").getAsString());
    } finally {
      appAndDb.app().stop();
      appAndDb.spicedb().close();
    }
  }

  @Test
  void bobDeniedWhenNotViewer() throws Exception {
    var appAndDb =
        Server.createApp(
            List.of(
                Server.rel("document:doc1", "folder", "folder:marketing"),
                Server.rel("folder:marketing", "viewer", "user:alice")));

    appAndDb.app().start(0);
    int port = appAndDb.app().port();

    try {
      var r = get(port, "/drive/marketing/doc1", "bob");
      assertEquals(403, r.statusCode());
      assertEquals("Forbidden", json(r.body()).get("error").getAsString());
    } finally {
      appAndDb.app().stop();
      appAndDb.spicedb().close();
    }
  }

  @Test
  void inheritsViewFromParentFolder() throws Exception {
    var appAndDb =
        Server.createApp(
            List.of(
                Server.rel("folder:marketing", "parent", "folder:root"),
                Server.rel("document:doc1", "folder", "folder:marketing"),
                Server.rel("folder:root", "viewer", "user:alice")));

    appAndDb.app().start(0);
    int port = appAndDb.app().port();

    try {
      var r = get(port, "/drive/marketing/doc1", "alice");
      assertEquals(200, r.statusCode());
      assertTrue(json(r.body()).get("ok").getAsBoolean());
    } finally {
      appAndDb.app().stop();
      appAndDb.spicedb().close();
    }
  }

  @Test
  void returns401WhenXUserMissing() throws Exception {
    var appAndDb =
        Server.createApp(
            List.of(
                Server.rel("document:doc1", "folder", "folder:marketing"),
                Server.rel("folder:marketing", "viewer", "user:alice")));

    appAndDb.app().start(0);
    int port = appAndDb.app().port();

    try {
      var r = get(port, "/drive/marketing/doc1", null);
      assertEquals(401, r.statusCode());
      assertEquals("Missing X-User header", json(r.body()).get("error").getAsString());
    } finally {
      appAndDb.app().stop();
      appAndDb.spicedb().close();
    }
  }
}
